Emotes

Sarahtonein opened the pull request:
Added functionality to detect contributors emojis. Reactions to these messages will occur, and respective contributor will be DM'd with a link to the message their emoji has been mentioned in. 

Contributors will NOT receive a DM if it is outside their UTC working hours. Timer is presently set to every 4 hours.

Balugrizzly commented on main.py:
```diff
@@ -0,0 +1,76 @@
+            if not contributor.actioned:
+                print(f'Messaging the user, {contributor.uid}')
+                message_link = message.jump_url  # Get the jump URL of the message
+                await send_dm_periodically(bot, contributor, message_link)
```
Why do you await this?
Won't this block?

Sarahtonein replied:
> Why do you await this? Won't this block?

I used await as this is a continuous task there are no blocks I can identify by my testing.
However after reviewing things I think I have found a potential block issue outside of this that needs to be checked

Balugrizzly reviewed (commented):
Lets not do the every 4h okey once is okey but i like the enthusiasm there haha

non persistent data on restart consciously?

Sarahtonein commented:
> Lets not do the every 4h okey once is okey but i like the enthusiasm there haha
> 
> non persistent data on restart consciously?

No worries I will remove the timer. 
I'll create a JSON file to store the contributor data

Sarahtonein commented:
Addressed change suggestions raised. Contributor data now stored in contributors.json.

Balugrizzly commented:
I did not see where you made the change to use persistent data but  i did not suggest that you do that i just asked but all good any way

Balugrizzly reviewed (commented):
Lookgs good
will have merge conflicts later

Sarahtonein commented:
> I did not see where you made the change to use persistent data but i did not suggest that you do that i just asked but all good any way

I added contributors.json to keep track of contributor information. This includes the name UID and emoji => uid mapping.

Your mention of persisting the data was a good idea and something. I was aware statically defining it in the code wasn't ideal but hadn't put much thought in it before your comment :)

Balugrizzly commented:
This is not what i meant by persistent data

The bot essentially keeps everything in memory which will therefore be bigger and if the memory goes all data in it goes 

one example would be:

You get mentioned on friday evening right after your official work hours 
now the bot essentially needs to hold everything necessary in memory until monday and with all the time zones etc.. this could be 3days or so in which it has to stay alive otherwise it will lose data and won't perfrom its intended functionality 

But like i said it was just a question

Sarahtonein commented:
> This is not what i meant by persistent data
> 
> The bot essentially keeps everything in memory which will therefore be bigger and if the memory goes all data in it goes
> 
> one example would be:
> 
> You get mentioned on friday evening right after your official work hours now the bot essentially needs to hold everything necessary in memory until monday and with all the time zones etc.. this could be 3days or so in which it has to stay alive otherwise it will lose data and won't perfrom its intended functionality
> 
> But like i said it was just a question

Thanks for the clarification :). For the time being I have removed the UTC hours thing and the bot simply sends a DM to the user once. I maybe misinterpreted what you initially said about not pinging the contributor every 4 hours and removed that part completely. can add this back in though if we want it easily.

Sarahtonein commented:
@Balugrizzly , have separated this from the combined branch and should now be ready to review

Balugrizzly commented on bot/commands.py:
```diff
@@ -51,6 +51,9 @@ def check(msg):
                 uid, emoji_id = inputs
 
+                # Strip the <@> characters from the uid
+                uid = uid.strip('<@!>')
```
You want to scrip "!" too?

Sarahtonein replied:
> You want to scrip "!" too?

Discord may include ! if the user has a nickname to my understanding so I wanted to strip ! if it is included as well.

Balugrizzly commented on emojis/emojis.py:
```diff
@@ -1,17 +1,23 @@
 from datetime import datetime
 from pytz import timezone
 
+# Get the directory of the current script
+script_dir = os.path.dirname(os.path.realpath(__file__))
+
+# Construct the path to the contributors.json file
+json_file_path = os.path.join(script_dir, 'contributors.json')
```
Not sure why you are doing this 
everything is static here or at least i would hope so

Sarahtonein replied:
Was having some problems with the file being read and duplicate file being created. Have added the file path as a constant and am using this instead.

Balugrizzly commented on README.md:
````diff
@@ -1 +1,88 @@
+```
+To add a contributor, please provide the following information:
+Name
+User ID (UID)
+Emoji ID
````
How are "normal" users getting the ID and the Emoji ID?
I can see that being very challenging and at a minimum laborious

Sarahtonein replied:
> How are "normal" users getting the ID and the Emoji ID? I can see that being very challenging and at a minimum laborious

Documentation yet to be updated after Lap provided feedback when testing the bot. Wording here was also slightly off. When adding a contributor you can provide the :emoji: and the bot will identify the Emoji ID.

Balugrizzly commented on bot/events.py:
```diff
@@ -0,0 +1,54 @@
+def setup_events(bot: commands.Bot):
+    @bot.event
+    async def on_message(message):
+        for emoji_id, contributor_uid in emoji_id_mapping.items():
+            contributor = next((c for c in contributors if c["uid"] == contributor_uid), None)
```
doubly neested loops are tight :)

simple for in would have done the trick just saying

Sarahtonein replied:
Nesting on nesting. I have changed this.

Balugrizzly commented on bot/events.py:
```diff
@@ -0,0 +1,54 @@
+                    except discord.errors.NotFound:
+                        logging.warning(f'User not found: {contributor["uid"]}')
+
+        if message.author == bot.user:
+            return
```
Would it make more sense to move this up?

Sarahtonein replied:
> Would it make more sense to move this up?

Totally would.

Balugrizzly commented on emojis/contributors.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "contributors": [
+    {"uid": "353572599957291010"},
+    {"uid": "368617749041381388"},
+    {"uid": "303732860265693185"},
+    {"uid": "406302927884386318"},
+    {"uid": "548974112131776522"},
+    {"uid": "154033306894073856"},
+    {"uid": "316765092689608706"}
```
Comments to which they belong would be nice or using constants so you don't have to cpy them twice

Sarahtonein replied:
> Comments to which they belong would be nice or using constants so you don't have to cpy them twice

Initially took a name parameter with the user ID and emoji. Was asked to change this. 

I have updated to add a note to each contributor which takes the name from the emoji so that we can identify who is who.

Balugrizzly reviewed (changes requested):
There are open points from other branches and from our discussions

Sarahtonein commented:
> There are open points from other branches and from our discussions

Yeah just a few :P I will work on these today

Balugrizzly commented:
Okey like i said in the other PR up to you how to proceed

Balugrizzly commented:
This still has merge conflicts

Sarahtonein commented:
> This still has merge conflicts

Conflicts now resolved :)
